### ISSUE TITLE
When Freehand is not finished, it is not available after cleaning

### ISSUE THREAD

**Issue by xhf127** [cornerstonejs/cornerstoneTools#306]
I have two questions. Thank your help

1.  I want to use a key remove the currently highlighted annotation，I find the active tool to remove it，but when I click Freehand to make it in active, I can't use Freehand after I delete it

````javascript
$(window).keydown(function (e) {
            if (e.keyCode !== 46) return;
            var element = $('.action').find('.imageViewerViewport')[0];
            var toolState = cornerstoneTools.globalImageIdSpecificToolStateManager.saveToolState();
            for (var image in toolState) {
                if (toolState.hasOwnProperty(image)) {
                    var toolData = toolState[image];
                    for (var toolType in toolData) {
                        if (toolData.hasOwnProperty(toolType)) {
                            var dataArray = toolData[toolType];
                            var data = dataArray['data'];
                            var removeData;
                            $.each(data, function (index, value) {
                                if (value.active === true) {
                                    removeData = value;
                                }
                            });
                            cornerstoneTools.removeToolState(element, toolType, removeData);
                        }
                    }
                }
            }
            cornerstone.updateImage(element);
        })
````

````javascript
Uncaught TypeError: Cannot set property 'active' of undefined
    at endDrawing (cornerstoneTools.js:6620)
    at HTMLDivElement.mouseDownCallback (cornerstoneTools.js:6656)
    at HTMLDivElement.dispatch (jquery-2.1.0.js:4371)
    at HTMLDivElement.elemData.handle (jquery-2.1.0.js:4057)
    at Object.trigger (jquery-2.1.0.js:4286)
    at HTMLDivElement.<anonymous> (jquery-2.1.0.js:4828)
    at Function.each (jquery-2.1.0.js:381)
    at jQuery.fn.init.each (jquery-2.1.0.js:137)
    at jQuery.fn.init.trigger (jquery-2.1.0.js:4827)
    at HTMLDivElement.mouseDown (cornerstoneTools.js:11046)`
````

2.  When I did not finish drawing  Freehand, I called the cleaning method, and I can't use the Freehand.

````javascript
var toolStateManager = cornerstoneTools.globalImageIdSpecificToolStateManager;
toolStateManager.clear(element);
````

**Comment by dannyrb** [comment]
So the issue here is that you cleared the tooldata for "freehand" in the middle of using it, but the tool expects that data to still exist? If I'm misunderstanding, please feel free to elaborate

**Comment by xhf127** [comment]
The tool thinks that the last drawing is not finished, so it is not to start drawing again. The currentTool in the free configuration is not less than 0, but data is no longer there. At free function mouseDownCallback (). In addition, there are drawHandlesOnHover in all the other tool configurations,  not in free。 I found the code， but how do I configure it
`
// If the tool is active, draw a handle at the cursor location
                    var options = {
                        fill: fillColor
                    };

                    if (data.active) {
                        (0, _drawHandles2.default)(context, eventData, config.mouseLocation.handles, color, options);
                    }
                    // Draw the handles
                     (0, _drawHandles2.default)(context, eventData, data.handles, color, options);`
 exports.default = function (context, renderData, handles, color, options) {
                context.strokeStyle = color;

                Object.keys(handles).forEach(function (name) {
                    var handle = handles[name];

                    if (handle.drawnIndependently === true) {
                        return;
                    }

                    if (options && options.drawHandlesIfActive === true && !handle.active) {
                        return;
                    }

                    context.beginPath();

                    if (handle.active) {
                        context.lineWidth = _toolStyle2.default.getActiveWidth();
                    } else {
                        context.lineWidth = _toolStyle2.default.getToolWidth();
                    }

                    var handleCanvasCoords = _externalModules2.default.cornerstone.pixelToCanvas(renderData.element, handle);

                    context.arc(handleCanvasCoords.x, handleCanvasCoords.y, handleRadius, 0, 2 * Math.PI);

                    if (options && options.fill) {
                        context.fillStyle = options.fill;
                        context.fill();
                    }

                    context.stroke();
                });
            };

**Comment by dannyrb** [comment]
It looks like you've pasted a snippet of transpiled/compiled code, and not the original source. My best guess is that you delete the data, but the event listeners setup by the tool are still in place. I don't know that you can configure it without modifying the source. Currently, the easiest way to extend an existing tool is to pull down the source for the [repository, modify it, and rebuild](https://github.com/cornerstonejs/cornerstoneTools#build-system).

It's worth noting, that if you fork the source, it is often better to create a new tool instead of overwriting an existing one. This makes merging updates in from our master easier down the road.

> If you're unfamiliar with our build process, or think the documentation could be better, please let me know. 

A _better_ long term solution would be for you to propose a way for us to modify each tool to allow extensions like you're proposing without the need to modify the code's original source.

**tl;dr**
- If you post the "original source" instead of the output, it'll be easier for me to help
- Better approaches:
    - A PR/discussion about adding support for handling key presses
    - A PR/discussion that gives users more control of extending functionality without touching the source

**Comment by xhf127** [comment]
The first problem I think can be avoided when drawing unfinished to delete data. The sticker code is the source code, I have intercepted two parts of the code, part of the drawHandles.js,the other part is https://github.com/cornerstonejs/cornerstoneTools/blob/954898e1f98f218690c36819363ffbf5b8e2fa30/src/imageTools/freehand.js#L386
I will temporarily modify the source code to achieve the second problem

**Comment by dannyrb** [comment]
Sounds good :+1:

I think your question has been answered? If you think the library should change to make accomplishing something easier, or that we should add a new feature (ie. keypress support), please feel free to make a separate issue. If your question _has not_ been answered, you're more than welcome to post a follow-up question and to re-open this issue.
